glommio/src/io/read_result.rs

```diff
@@ -2,104 +2,138 @@
-/// ReadResult encapsulates a buffer, returned by read operations like [`get_buffer_aligned`] and
-/// [`read_at`]
-///
-/// [`get_buffer_aligned`]: struct.DmaStreamReader.html#method.get_buffer_aligned
-/// [`read_at`]: struct.DmaFile.html#method.read_at
+#[derive(Clone, Debug)]
+/// ReadResult encapsulates a buffer, returned by read operations like
+/// [`get_buffer_aligned`](super::DmaStreamReader::get_buffer_aligned) and
+/// [`read_at`](super::DmaFile::read_at)
 pub struct ReadResult {
-    buffer: Option<Rc<DmaBuffer>>,
+    inner: Option<ReadResultInner>,
+}
+
+impl core::ops::Deref for ReadResult {
+    type Target = [u8];
+
+    /// Allows accessing the contents of this buffer as a byte slice
+    fn deref(&self) -> &[u8] {
+        self.inner.as_deref().unwrap_or(&[])
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ReadResultInner {
+    buffer: Rc<DmaBuffer>,
     offset: usize,
-    end: usize,
+    len: NonZeroUsize,
 }
 
-static EMPTY: [u8; 0] = [0; 0];
+impl core::ops::Deref for ReadResultInner {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        &self.buffer.as_bytes()[self.offset..][..self.len.get()]
+    }
+}
 
-#[allow(clippy::len_without_is_empty)]
 impl ReadResult {
-    pub(crate) fn empty_buffer() -> ReadResult {
-        ReadResult {
-            end: 0,
-            offset: 0,
-            buffer: None,
-        }
+    pub(crate) fn empty_buffer() -> Self {
+        Self { inner: None }
     }
 
-    pub(crate) fn from_whole_buffer(buffer: DmaBuffer) -> ReadResult {
-        ReadResult {
-            end: buffer.len(),
-            offset: 0,
-            buffer: Some(Rc::new(buffer)),
+    pub(crate) fn from_whole_buffer(buffer: DmaBuffer) -> Self {
+        Self {
+            inner: NonZeroUsize::new(buffer.len()).map(|len| ReadResultInner {
+                buffer: Rc::new(buffer),
+                offset: 0,
+                len,
+            }),
         }
     }
 
-    /// Copies the contents of this [`ReadResult`] into the byte-slice `dst`.
-    ///
-    /// The copy starts at position `offset` into the [`ReadResult`] and copies
-    /// as many bytes as possible until either we don't have more bytes to copy
-    /// or `dst` doesn't have more space to hold them.
+    /// Creates a slice of this ReadResult with the given offset and length.
     ///
-    /// [`ReadResult`]: struct.ReadResult.html
-    pub fn read_at(&self, offset: usize, dst: &mut [u8]) -> usize {
-        let offset = self.offset + offset;
-        self.buffer.as_ref().unwrap().read_at(offset, dst)
+    /// Returns `None` if either offset or offset + len would not fit in the original buffer.
+    pub fn slice(this: &Self, extra_offset: usize, len: usize) -> Option<Self> {
+        if let Some(len) = NonZeroUsize::new(len) {
+            ReadResultInner::slice(this.inner.as_ref()?, extra_offset, len)
+                .map(|inner| Self { inner: Some(inner) })
+        } else {
+            Some(Self::empty_buffer())
```

Comment:
This is needed to make sure that calls to `slice` with `len = 0` are valid. If they aren't valid, then `len` should be changed to `NonZeroUsize`.